File: interop/go/kkrpc/protocol.go
package kkrpc

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"time"
)

const CallbackPrefix = "__callback__"

func init() {
	rand.Seed(time.Now().UnixNano())
}

func GenerateUUID() string {
	parts := make([]string, 0, 4)
	for i := 0; i < 4; i++ {
		parts = append(parts, fmt.Sprintf("%x", rand.Int63()))
	}
	return fmt.Sprintf("%s-%s-%s-%s", parts[0], parts[1], parts[2], parts[3])
}

func EncodeMessage(payload map[string]any) (string, error) {
	data, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	return string(data) + "\n", nil
}

func DecodeMessage(raw string) (map[string]any, error) {
	var payload map[string]any
	if err := json.Unmarshal([]byte(raw), &payload); err != nil {
		return nil, err
	}
	return payload, nil
}

File: interop/python/kkrpc/protocol.py
import json
import random
from typing import Any, Dict

ARG_ENVELOPE_TAG = "__kkrpc_next_arg__"


def generate_uuid() -> str:
	return "-".join(f"{random.getrandbits(53):x}" for _ in range(4))


def encode_message(payload: Dict[str, Any]) -> str:
	return json.dumps(payload, ensure_ascii=False) + "\n"


def decode_message(message: str) -> Dict[str, Any]:
	return json.loads(message)
